### DISCUSSION TITLE
Event timing strange behaviour

### DISCUSSION THREAD

**Discussion by craig-warren** [pyopencl/pyopencl#543] [Troubleshooting]
Hi,

I'm getting some very strange values when trying to time events. The results are orders of magnitude different from expected. The is an example of a function which is called repeatedly in a for loop:
```
def update_magnetic(self):
        """Update magnetic field components."""
        event = self.update_magnetic_dev(np.int32(self.grid.nx),
                                         np.int32(self.grid.ny),
                                         np.int32(self.grid.nz),
                                         self.grid.ID_dev,
                                         self.grid.Hx_dev,
                                         self.grid.Hy_dev,
                                         self.grid.Hz_dev,
                                         self.grid.Ex_dev,
                                         self.grid.Ey_dev,
                                         self.grid.Ez_dev)
        event.wait()
        self.compute_time += (event.profile.end - event.profile.start)*1e-9
```
Multiple functions can update `self.compute_time` but every function has a `event.wait()`. When I print `self.compute_time` at the end it is not what is expected. I'm I doing something stupid here?

Thanks,
Craig

**Comment by inducer** [comment]
This looks fine from a quick look.

- What ICD is this?
- Is your queue in profiling mode?

**Reply by craig-warren** [reply]
Thanks @inducer. 

- Profiling is enabled on the queue. 
- How can I check what ICD is being used? I'm using macOS Monterey (12.2) on a Mac Pro with an AMD Radeon RX580 GPU.

**Reply by inducer** [reply]
> How can I check what ICD is being used?

`clinfo`, or by printing properties of `Device`, or by `examples/dump-properties.py`.

If kernel event profiling is broken on that ICD (it very well might be, the Apple ICDs don't have a stellar reputation), maybe marker-based timing can work? https://documen.tician.de/pyopencl/runtime_queue.html?highlight=enqueue_marker#pyopencl.enqueue_marker

**Reply by craig-warren** [reply]
Thanks. Definitely Apple ICD being used.

Can you explain a bit more how I could try the marker-based timing?

Do you think its worth me trying the `pocl` ICD for CPU? - https://documen.tician.de/pyopencl/misc.html#enabling-access-to-cpus-and-gpus-via-py-opencl

**Reply by inducer** [reply]
> Can you explain a bit more how I could try the marker-based timing?

Sure: enqueue a marker before what you'd like to time, enqueue a marker after, wait on both events, look at the time stamp difference to obtain timing.

> Do you think its worth me trying the pocl ICD for CPU?

I'm confident it'll *work*. Folks around me use it all the time. But it'll likely not go much faster than CPU C code. Don't expect it to magically vectorize without substantial help.

**Reply by craig-warren** [reply]
Marker-based timing appears to work with the Apple ICD on CPU but not GPU - it shows zero time.

I haven't been able to get the pocl ICD working yet. It will install with conda but isn't picked up yet, through either clinfo or pyopencl. Any thoughts?
